### ISSUE TITLE
grpo训练结果不正常（completions/clipped_ratio 长期为1.0）

### ISSUE THREAD

**Issue by XuJH5080** [modelscope/ms-swift#4950]
尊敬的老师您好：

首先提供我的脚本：

![Image](https://github.com/user-attachments/assets/c0c13691-eda1-41db-9c79-9a28c1bb32e2)

CUDA VISIELE DEVICES=4 \
MAX PIXELS=1048576 
PYTORCH CUDA ALLOC CONF='expandable segments:True' \
swift rIhf 
--rihf type grpo \
--model /home/zmk/xjh/qiyuan/baseline_vit sft 07131103/v0-20250713-110853/checkpoint-800（baseline是qwen2.5-vl-7b）
--external _plugins examples/train/grpo/plugin/plugin.py 
--reward funcs rouge reward sbert semantic similarity \
--reward weights 0.7 0.3 \
--train type lora \
--lora rank 8 \
--lora alpha 32 \
--target modules all-linear \
-torch dtype bfloat16 \
--dataset "/home/zmk/xih/dataset /fusai dataset/rematch example 0613/eval-dataset-example/ground truth" \
--max completion length 512 \
--num train epochs 1 \
--per device train batch size 4 \
--per device eval batch size 4 \
--learning rate 1e-5 \
--gradient accumulation steps 1 
--eval steps 25 \
--save_steps 25 \
--save total limit 2 \
--logging_steps 3 \
--max_length 3070 \
--truncation strategy delete \
--max_pixels 1048576 \
--output_dir /home/zmk/xjh/giyuan/output/test-grpo_87141234 \
--warmup_retio 0.05 \
--dataloader_num_workers 4 \
--dateset_num_proc 4 \
--num_generations 4 \
--temperature 0,7 \ 
--system 'examples/trein/grpo/prompt,giyuan.txt' \
--1og completions true \
--split_dataset_ratio 0 \
--freeze_vit True 

其次是运行时的输出：（硬件就是单卡a100）

![Image](https://github.com/user-attachments/assets/ee5e0fed-d6b5-459d-ba26-46b7019fd57e)

{'loss’: 0.00292034, "grad norm': 2.98987532, 'learning rate’: 1e-08, 'memory(GiB)': 40.11, 'train speed(iter/s)": 0.032308,completions/mean length’: 17.41666667, 'completions/min length': 16.666667, 'completions/max length': 18.0,"completions/ciipped ratio’: 1.0, 'rewards/RougeRewardFunction/mean’: 0.0, 'rewards/RougeRewardFunction/std’: 0.0, 'rewards/SentenceBERT RewardFunction/mean’: 0.52444987,'rewards/sentenceBERT RewardFunction/std': 0.05381423,'reward': 0.15733497, 'reward std*: 0.01614427, "K1': 0.07304342, 'clip ratio/low mean': 0.0, 'clip ratio/low min': 0.0, 'clip ratio/high mean': 0.0, "clip ratio/high nax': 0.0. *clip ratio/reeion mean': 0.0, 'epoch': 0.98, 'global step/max steps’: '672/685',"percentage’: "98.10%', 'elapsed time':"5h 46m 39s’,"remaining time': '6m 42s'}

可以看到completions/mean_length等长度相关的几个指标远远达不到512，但是截断比率长期为1.0，这让我非常费解，而且整个训练中的奖励并没有出现过上升的趋势，这表明似乎真的没有训练（因为所有的回答都因为超长被舍弃了）。

谢谢大佬帮助，祝好！

**Comment by hjh0119** [comment]
can you set `--report_to wandb/swanlab` to see the content of prompts and completions? (swanlab supported in main branch)

**Comment by XuJH5080** [comment]
<img width="2712" height="1140" alt="Image" src="https://github.com/user-attachments/assets/2d24be24-e232-4e82-b415-cdff404a1db4" />

<img width="2715" height="1132" alt="Image" src="https://github.com/user-attachments/assets/896a7307-841e-42f6-a39f-225d9e116773" />

<img width="2726" height="1123" alt="Image" src="https://github.com/user-attachments/assets/30074f08-14e3-4877-af95-aff38bf481f1" />

<img width="2698" height="1123" alt="Image" src="https://github.com/user-attachments/assets/5eebaeda-eb42-47b8-a33d-432805b269d6" />

大佬您好，以上是wandb的截图，图中也能明显看到我描述的问题，就是length都没到512，截断比率一直是1.0

**Comment by hjh0119** [comment]
I mean the completion table. Anything unusual? Or do you have any scripts I can use to reproduce this issue?

**Comment by XuJH5080** [comment]
> I mean the completion table. Anything unusual? Or do you have any scripts I can use to reproduce this issue?

感谢您的回复，我是在一个离线服务器上跑的grpo，wandb没能上传成功completion table。我已经在拉取main分支的代码，重新构建程序环境，用来复现错误。

至于您说的复现这个issue，在我的实验中，这个现象一直存在，可能和我使用的数据集等有关，可能不太容易在您那里复现，但是我建议大佬您可以尝试一下加上参数“--truncation strategy delete”看看能否复现我的错误。因为我参考了其他的帖子，其中有说过grpo不要使用--truncation_strategy left/right（我找了很久，没有重新找到那个issue的回复），但是在我添加了这个--truncation_strategy delete参数后，遇到了一个assert报错，指定--truncation_strategy必须为left，但因为那个帖子的缘故，我当时直接修改了源码，将left设置成了delete，我也会进行更多的尝试，希望您可以尝试一下这样能否复现报错（我用的是ms-swift==3.5.1 从源码以可编辑依赖形式安装的）

我也会在用swanlab记录后，将结果继续记录在这个问题当中。

再次感谢大佬的回复，祝好！

**Comment by hjh0119** [comment]
The truncation strategy `delete` is not currently supported in GRPO, so you may encounter strange errors. My suggestion is to increase the length-related parameters to avoid training interruptions caused by over-length data.

**Comment by SONG0257** [comment]
--可以看到completions/mean_length等长度相关的几个指标远远达不到512，但是截断比率长期为1.0，这让我非常费解

I encountered the same problem when do GRPO training. the completions/clipped_ratio is always 1.0, but the reported completion lengths do not reach the --max_completion_length

**Comment by hjh0119** [comment]
@SONG0257 any repro scripts?

**Comment by hjh0119** [comment]
delete strategy is supported in https://github.com/modelscope/ms-swift/pull/4977

**Comment by lizhyuxi** [comment]
你这里的 clip_ratio 不是1而是0啊，代表clip根本没生效，看了下你的脚本，你没有设置 num_iterations 参数，该参数默认值是1，此时 clip 策略不生效，所以 clip_ratio恒等于0是符合预期的，至于为什么该情况clip不生效，你可以看下这个 PR：https://github.com/huggingface/trl/discussions/3548

**Comment by XuJH5080** [comment]
> 你这里的 clip_ratio 不是1而是0啊，代表clip根本没生效，看了下你的脚本，你没有设置 num_iterations 参数，该参数默认值是1，此时 clip 策略不生效，所以 clip_ratio恒等于0是符合预期的，至于为什么该情况clip不生效，你可以看下这个 PR：[huggingface/trl#3548](https://github.com/huggingface/trl/discussions/3548)

首先，请看wandb的倒数第三行的第三张图，clip ratio的确为1，只有其他相关的参数为什么是0，我就也搞不清楚了，之前因为一些事情前一个实验被终止了，目前我还没有再次遇到这个问题，如果遇到了我会接着评论。最后感谢您分享的这个帖子，解决了“为什么起始loss为0”这个困惑，Thanks♪(･ω･)ﾉ

**Comment by zerzerzerz** [comment]
> 你这里的 clip_ratio 不是1而是0啊，代表clip根本没生效，看了下你的脚本，你没有设置 num_iterations 参数，该参数默认值是1，此时 clip 策略不生效，所以 clip_ratio恒等于0是符合预期的，至于为什么该情况clip不生效，你可以看下这个 PR：[huggingface/trl#3548](https://github.com/huggingface/trl/discussions/3548)

太感谢了老哥，我训练的时候也发现train/clip_ratio/high_max等一共6个相关的metrics一直是0，还纳闷为什么会这样，看了看num_iterations = 1，豁然开朗啊。太感谢了！

**Comment by lizhyuxi** [comment]
> > 你这里的 clip_ratio 不是1而是0啊，代表clip根本没生效，看了下你的脚本，你没有设置 num_iterations 参数，该参数默认值是1，此时 clip 策略不生效，所以 clip_ratio恒等于0是符合预期的，至于为什么该情况clip不生效，你可以看下这个 PR：[huggingface/trl#3548](https://github.com/huggingface/trl/discussions/3548)
> 
> 首先，请看wandb的倒数第三行的第三张图，clip ratio的确为1，只有其他相关的参数为什么是0，我就也搞不清楚了，之前因为一些事情前一个实验被终止了，目前我还没有再次遇到这个问题，如果遇到了我会接着评论。最后感谢您分享的这个帖子，解决了“为什么起始loss为0”这个困惑，Thanks♪(･ω･)ﾉ

不好意思之前忽略了这张图，这样看来确实很奇怪，尤其是你下面所有相关参数 low_mean, high_mean 都是 0，这代表其本身就是不自洽的，除非你的 rollout response 确实都是空回复（这也不正常），况且 num_iterations=1 应该就是完全关闭 clip 才对...
可能确实是现版本 swift 的 bug，如果能稳定复现建议可以提个 PR
